### PR TITLE
Purpose new FIP for address/domain deactivation ( FIP #7 )

### PR DESCRIPTION
**Pull request by dapixcasey**

### CODE REVIEW

File: fip-6.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+##### Response
+##### Exception handling
+|Error condition|Trigger|Type|fields:name|fields:value|Error message|
+|---|---|---|---|---|---|
+|Invalid FIO Address format|FIO Address format is not valid|400|"fio_address"|Value sent in, e.g. "purse@alice"|"Invalid FIO Address"|
+|Not owner of FIO Address|The signer does not own the address|403||||
+|Invalid fee value|max_fee format is not valid|400|"max_fee"|Value sent in, e.g. "-100"|"Invalid fee value"|
+|Insufficient funds to cover fee|Account does not have enough funds to cover fee|400|"max_fee"|Value sent in, e.g. "400000000"|"Insufficient funds to cover fee"|
+|Fee exceeds maximum|Actual fee is greater than supplied max_fee|400|max_fee"|Value sent in, e.g. "400000000"|"Fee exceeds supplied maximum"|
+|Invalid TPID|TPID is not empty or contains invalid FIO address|400|"tpid"|Value sent in is not empty and not a valid FIO Address format|"TPID must be empty or valid FIO address"|
```
Comment (lukestokes) [inline fip-6.md:57]: `empty or contains` should probably be `empty and contains`

File: fip-6.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+{
+    "status": "Ok",
+    "fee_collected": "800000000"        
+}
+```
+## Fees
+A new fee will be created for `deactivate_fio_domain` . This fee will be of type 0 and should cost 800000000 SUF.
+
+## Rationale
+Some users might find it useful to have the ability to deactivate unused addresses or domains. This feature proposal will allow Address owners the ability to burn their addresses at any time. Domain owners that are looking to deactivate will have their expiration set to the current time. Setting this expiration allows for the continued use of the domain during a grace period until the new owner registers. If the domain does not receive a new owner in time, all FIO Addresses linked to that domain will be burned after their expirations but will not be able to be utilized.
```
Comment (lukestokes) [inline fip-6.md:130]: What is the grace period? Does the grace period start immediately, regardless of how far into the future the domain was paid up?

As to burning all FIO addresses, can you outline a bit more on how the grace period impacts them, how they will know the domain they are on needs to be paid for, etc? What happens when someone buys up the domain and it's transferred to a new owner (nothing?)? If an individual address user can pay the fee for the domain in order to keep their address functional, that would be good to highlight here. Also, people may be unwilling to pay for addresses if they can't know for sure if the domain will be around in the future.

File: fip-6.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+## Abstract
+This FIP implements the following:
+* Adds the ability for FIO Address owners the ability to deactivate stale or unused addresses.
+* Adds the ability for FIO Domain owners the ability to set their domain as expired.
+* Adds new API end points for FIO Domain deactivation and FIO Address burn. 
+* Adds new fees for FIO Domain deactivation and FIO Address burn.
+
+## Motivation
+Presently the FIO Blockchain only burns and removes old addresses and domains once they reach their expiration date. Added functionality to remove these addresses and domains at any given time gives the owners flexibility.
```
Comment (lukestokes) [inline fip-6.md:19]: Does this in any way impact the ability of domain owners to control the addresses registered on their domain? As in, can a domain owner burn an address on their domain owned by someone else?

File: fip-6.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+created: 2020-04-22
+updated: 
+---
+
+## Abstract
+This FIP implements the following:
+* Adds the ability for FIO Address owners the ability to deactivate stale or unused addresses.
+* Adds the ability for FIO Domain owners the ability to set their domain as expired.
+* Adds new API end points for FIO Domain deactivation and FIO Address burn. 
+* Adds new fees for FIO Domain deactivation and FIO Address burn.
```
Comment (pawmmm) [inline fip-6.md:16]: So is it burn or deactivate. I feel like deactivate implies it will be around and can be reactivated.

File: fip-6.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+## Motivation
```
Comment (pawmmm) [inline fip-6.md:18]: Should add motivation for "ability to set their domain as expired". I'm not sure what is the reason to have that beyond "more flexibility".

File: fip-6.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+Presently the FIO Blockchain only burns and removes old addresses and domains once they reach their expiration date. Added functionality to remove these addresses and domains at any given time gives the owners flexibility.
+
+## Specification
+### Burn Addresses
+#### New end point: *deactivate_fio_address*
```
Comment (pawmmm) [inline fip-6.md:23]: Burn or deactivate?

File: fip-6.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+|fee_collected|String|fee amount collected SUFs|
+###### Example
+```
+{
+    "status": "Ok",
+    "fee_collected": "400000000"		
+}
+```
+###### New fee
+A new fee will be created for `deactivate_fio_address` . This fee will be of type 0 and should cost 400000000 SUF.
```
Comment (pawmmm) [inline fip-6.md:75]: Is it bundle eligible? What's fee type 0?

File: fip-6.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+|fee_collected|String|fee amount collected SUFs|
+###### Example
+```
+{
+    "status": "Ok",
+    "fee_collected": "800000000"        
+}
+```
+## Fees
+A new fee will be created for `deactivate_fio_domain` . This fee will be of type 0 and should cost 800000000 SUF.
```
Comment (pawmmm) [inline fip-6.md:127]: Is it bundle eligible? What's fee type 0?

File: fip-6.md (hunk 2 above)
Comment (pawmmm) [inline fip-6.md:130]: Per FIP template "The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made." This section belongs in Motivation.

File: fip-6.md (hunk 1 above)
Comment (dapixcasey) [reply fip-6.md:57]: "TPID must be empty or valid FIO address" is the standard error across all associated endpoints. In software, something cannot be empty AND contain data.

File: fip-6.md (hunk 2 above)
Comment (dapixcasey) [reply fip-6.md:130]: Currently, `DOMAINWAITFORBURNDAYS` and `ADDRESSWAITFORBURNDAYS` are set inside fio.common at 90 days.

File: fip-6.md (hunk 3 above)
Comment (dapixcasey) [reply fip-6.md:19]: Domain owners should not have any control over the addresses registered on their domain. There is no parent/child relationship.

File: fip-6.md (hunk 4 above)
Comment (dapixcasey) [reply fip-6.md:16]: I agree.  I will make the changes to better describe what is actually happeneing

File: fip-6.md (hunk 6 above)
Comment (dapixcasey) [reply fip-6.md:23]: Addresses are not being burned. Just slotted to be burned when burnexpired is called.

File: fip-6.md (hunk 7 above)
Comment (dapixcasey) [reply fip-6.md:75]: Took this style from Eds FIPs but I can be more descriptive here.

File: fip-6.md (hunk 1 above)
Comment (dapixcasey) [reply fip-6.md]: I was using a reference from Eds FIPs but I've changed it to match the current  1.0 API spec

File: fip-6.md (hunk 4 above)
Comment (dapixcasey) [reply fip-6.md]: Technically, if they renew their domain or address during the 90 grace period, they are essentially reactivating it, yes?

**Comment by dapixcasey** [comment]
Should TPID rewards be removed? Thoughts?

**Comment by dapixcasey** [comment]
> **FIO Address**
> 
>     * It would be nice to have the contract action name match up with the endpoint name. Either use burn_fio_address or change action to deactivateaddress?
> 
>     * Needs more information on all of the other FIO functionality this will impact. What happens to FIO requests with that address? What if that address is listed as a proxy? What about the voters table? I recommend going through every action and table that uses a FIO Address and detail the changes that will need to be made.
> 
>     * Remove tpid award?
> 
> 
> FIO Domain
> 
>     * Match up the contract action with the API endpoint. I recommend "expire_fio_domain" since that is what is really happening.
> 
>     * Remove tpid award?

Renamed relevant calls as well as added information of effected tables, etc. Functionality to domain deactivation will be similar to expired logic. Owners are able to renew during the buffer period for continued use. I tend to agree with the thinking that tpid rewards should not apply.

File: fip-6.md (hunk 2 above)
Comment (ericbutz) [reply fip-6.md]: @lukestokes interesting idea on the ability to pay the Domain fee before it is burned (but after it is in the expired grace period) but that could be strange as someone could take over the hosting of the domain your address is on without you knowing it. As an FYI, the full details on domain expiration are at: https://stealth.atlassian.net/wiki/spaces/DEV/pages/90898460/FIO+Protocol#FIOProtocol-ExpiredDomain
I think the idea is that certain address and domain functions will stop working when a domain is expired and that will serve as notification that something is wrong (they will return an error "400:FIO Domain expired").

Comment (dapixcasey) [reply fip-6.md]: Could help to change the effected calls to return "400: FIO Domain expired or deactivated"

**Pull request merged**